### DISCUSSION TITLE
Are there any patterns for splitting a configuration into multiple files?

### DISCUSSION THREAD

**Discussion by MarkSRobinson** [apple/pkl#421] [Q&A]
I'm trying to puzzle my way through porting a helm chart to PKL. In Helm, it's very common to split the whole configuration into multiple files for readability purposes.

For a simple Helm chart, it would be pretty easy to squish all the objects into a single file and distribute that package and let people amend it. So my local config would amend the upstream file.

However, if I want to break apart my configuration to make it easier to manage into multiple files that would then require all users to create an amending module for each one of my files. I don't like this as it would effectively require all consumers of my chart to be exposed to the internal structure.

A somewhat more user-friendly way to do this would be to create an `amend` chain, so A amends B, amends C, amends D. Then the end-user would just need to `amends "A"` to pull in the full item. It's better from a user-perspective but feels really awkward to build a long amends chain.

Are there any other patterns for composing configuration over multiple files?

**Comment by HT154** [comment]
One very common pattern is to use [glob imports](https://pkl-lang.org/main/current/language-reference/index.html).

It's a contrived example, but you might have a file structure like this:

```
all.pkl
configs/
  deployment.pkl
  service.pkl
  ingress.pkl
```

And `all.pkl` might look like this, using pkl-k8s (imports omitted for brevity):
```pkl
resources: Listing<K8sResource> = new {
  for (_, resource in import*("configs/*.pkl")) {
    resource
  }
}

output {
  renderer = new YamlRenderer {}
  value = resources
}
```

**Comment by holzensp** [comment]
@HT154's example is totally valid. Also, always remember that users can arbitrarily `import` stuff they want also, so if your template is monolithic, they could still
```
resources = import("myResources.pkl")
```

Depending on what sort of config you're rolling, if it's "feature-based," you can also use the _presence_ of certain files to signal a feature is being used, so
```
app.pkl
redis.pkl
postgres.pkl
apps/
  someAppThatOnlyUsesPostgres/
    app.pkl
    postgres.pkl
  someAppThatOnlyUsesRedis/
    app.pkl
    redis.pkl
  someAppThatUsesEverything/
    app.pkl
    postgres.pkl
    redis.pkl
```

It depends very much on your wider context which patterns to pick.
